DB-2936 modify our helm chart to deploy VerticalPodAutoscaler

rluidash opened the pull request:
Deploy vertical pod autoscaler from dashbase helm chart

Based on the unofficial vpa helm chart from https://github.com/cowboysysop/charts/tree/master/charts/vertical-pod-autoscaler.

VPA is called on as dependency  when `vertical-pod-autoscaler` is made  `enabled: true` in dashbase-values.yaml file. And VPA components are deployed in dashbase namespace.  By default , the dashbase helm chart values.yaml file is set to false for `vertical-pod-autoscaler`

```
vertical-pod-autoscaler:
  enabled: true
```

modification on unofficial chart includes:
- vpa admission controller TLS certificate expiration us set to 3650 days
- updated the README.md file for this dependency chart
- fixed resource limit on each vpa components based on official vpa yaml files.
   
   default resource limits for vpa pods
```
      admissionController
        resources:
          limits:
            cpu: 200m
            memory: 500Mi
          requests:
            cpu: 50m
            memory: 200Mi

      recommender
        resources:
          limits:
            cpu: 200m
            memory: 1000Mi
          requests:
            cpu: 50m
            memory: 500Mi

        updater
          resources:
            limits:
              cpu: 200m
              memory: 1000Mi
            requests:
              cpu: 50m
              memory: 500Mi
```

Dependency for this VPA to work
- K8s 1.16+
- Metrics Server 0.2+
- Helm 3.0+

Tested the dashbase install with dependency  chart `vertical-pod-autoscaler` , and three vpa pods are created.
![image](https://user-images.githubusercontent.com/56131833/101706364-c35cff80-3a3d-11eb-9908-5c6ed8ebfb79.png)

This `vertical-pod-autoscaler` helm chart will create the following K8s resources, and some of these vpa K8s resource names are prefixed with helm release name and is `dashbase`

 secrets
* dashbase-vertical-pod-autoscaler-admission-controller-tls
* dashbase-vertical-pod-autoscaler-admission-controller-tokeXXXXX
* dashbase-vertical-pod-autoscaler-recommender-token-XXXXX
* dashbase-vertical-pod-autoscaler-updater-token-XXXXX

pods
* dashbase-vertical-pod-autoscaler-admission-controller-XXXXXXXX
* dashbase-vertical-pod-autoscaler-recommender-XXXXXXXXX-XXXXX
* dashbase-vertical-pod-autoscaler-updater-XXXXXXXXX-XXXXX

deployments
* dashbase-vertical-pod-autoscaler-admission-controller
* dashbase-vertical-pod-autoscaler-recommender
* dashbase-vertical-pod-autoscaler-updater

services
* dashbase-vertical-pod-autoscaler-admission-controller-metrics
* dashbase-vertical-pod-autoscaler-recommender-metrics
* dashbase-vertical-pod-autoscaler-updater-metrics
* vpa-webhook

mutatingwebhookconfiguration
* vpa-webhook-config



Testing with official vpa github `[example/hamster.yaml](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/examples/hamster.yaml)` file.. Based on [VPA testing described on here](https://docs.aws.amazon.com/eks/latest/userguide/vertical-pod-autoscaler.html) 

original hamster pod resource set on hamster.yaml
```
        resources:
          requests:
            cpu: 100m
            memory: 50Mi
```

After VPA recommender checked hamster pods and hamster-vpa  and adjust the resource accordingly
![image](https://user-images.githubusercontent.com/56131833/101706137-58abc400-3a3d-11eb-9691-5b0200efb75e.png)

Testing with embedded helm tests in this vertical-pod-autoscaler dependency chart templates/tests folder

![image](https://user-images.githubusercontent.com/56131833/101707094-6104fe80-3a3f-11eb-9eaa-41aa3655186f.png)

keitaf commented:
Thanks @rluidash !

Instead of modifying the VPA helm chart, can we try to directly use the VPA helm chart by doing the following?
- specify the cpu/memory limits in `admissionController.resources`, `recommender.resources` and `updater.resources` under `vertial-pod-autoscaler` in `dashbase/values.yaml`.
- generate a custom TLS cert with 3560 expiration day in our helm chart, and use it in the VPA chart using `admissionController.tls.existingSecret`

something like this:
`dashbase/values.yaml`
```
vertical-pod-autoscaler:
  enabled: false
  admissionController:
    resources:
          limits:
            cpu: 200m
            memory: 500Mi
    tls:
      existingSecret: {name of the secret}      
...
```
`dashbase/templates/vpa_tsl_secret.yaml` -> create a template to call `genCA` if VPA is enabled.

rluidash commented:
thanks for the input, making changes..

rluidash commented:
Created the vpa_tls_secret.yaml file  in dashbase/templates/ folder .. and this template is used for the vpa admission controller when vpa is enabled.

And  vpa_tls_secret.yaml file will create a secret named `<release>-vertical-pod-autoscaler-admission-controller-tls`
And this vpa tls secret contains: ca.crt, tls.crt, and tls.key  And the CN for issuer, CA and the TLS cert are below

```
ca.crt
  Issuer: CN = dashbase-vertical-pod-autoscaler-admission-controller
  Subject: CN = dashbase-vertical-pod-autoscaler-admission-controller

tls.crt
  Issuer: CN = dashbase-vertical-pod-autoscaler-admission-controller
  Subject: CN = vpa-webhook.dashbase.svc
```

Removed the original tls_secret.yaml file in the dependency chart `vertical-pod-autoscaler` `admissionController `template folder.

Update default dashbase values.yaml file  in such way that we can modify vpa components resources, replicas and future image version.

```
vertical-pod-autoscaler:
  enabled: false
  admissionController:
    image: "k8s.gcr.io/autoscaling/vpa-admission-controller:0.9.0"
    imagePullPolicy: IfNotPresent
    replicaCount: 1
    resources:
      limits:
        cpu: 200m
        memory: 500Mi
      requests:
        cpu: 50m
        memory: 200Mi
    tls:
      caCert: ""
      cert: ""
      key: ""
      existingSecret: ""
  recommender:
    image: "k8s.gcr.io/autoscaling/vpa-recommender:0.9.0"
    imagePullPolicy: IfNotPresent
    replicaCount: 1
    resources:
      limits:
        cpu: 200m
        memory: 1000Mi
      requests:
        cpu: 50m
        memory: 500Mi
  updater:
    image: "k8s.gcr.io/autoscaling/vpa-updater:0.9.0"
    imagePullPolicy: IfNotPresent
    replicaCount: 1
    resources:
      limits:
        cpu: 200m
        memory: 1000Mi
      requests:
        cpu: 50m
        memory: 500Mi
```

Tested  the changes  fresh install without previous vpa_tls_secert, and  existing vpa secret. Attached with helm install output (contains rendered manifests) for both cases   (see below files)
[helm-install-output-existing-vpa-tls-12142020f.txt](https://github.com/dashbase/dashbase-k8s/files/5692516/helm-install-output-existing-vpa-tls-12142020f.txt)
[helm-install-output-new-vpa-tls-12142020e.txt](https://github.com/dashbase/dashbase-k8s/files/5692518/helm-install-output-new-vpa-tls-12142020e.txt)

Tested with official examples `hamster.yaml` file for `hasmter-vpa`  and result is positive after this change


```
  Recommendation:
    Container Recommendations:
      Container Name:  hamster
      Lower Bound:
        Cpu:     525m
        Memory:  262144k
      Target:
        Cpu:     587m
        Memory:  262144k
      Uncapped Target:
        Cpu:     587m
        Memory:  262144k
      Upper Bound:
        Cpu:     1
        Memory:  500Mi
```

keitaf commented on dashbase/Chart.yaml:
```diff
@@ -21,3 +21,7 @@ dependencies:
   condition: metrics-server.enabled
+- name: vertical-pod-autoscaler
+  version: 0.9.0 # chart version is match with vpa image version
+  repository: "file://../vertical-pod-autoscaler"
```
Now certs and resources can be created/configured in our helm chart, can we just depend on the VPA chart and remove the copy&pasted files under `vertical-pod-autoscaler`?
```
- name: vertical-pod-autoscaler
  version: 0.9.0
  repository: https://cowboysysop.github.io/charts/
  condition: vertical-pod-autoscaler.enabled
```

rluidash replied:
I think if dashbase chart is built; it has the stuff from `file://../vertical-pod-autoscaler`
 Are you saying that use the vpa chart from repository https://cowboy.github.io/charts  like the metric-server does?

There are some difference or modification  between file://../vertical-pod-autoscaler  vs  https://cowboysysop.github.io/charts/ 

And originally in the previous PR changes,  we have the admission controller TLS cert  created from the dependency `vpa chart` ; then we just pass the Values from parent chart to subchart. However, the recent PR changes moved the  vpa admission controller TLS cert from dependency chart to parent chart. This changes the VPA subcart's `_helper.tpl  file, and removed the orginal tls_cert.yaml file in the vpa subcart. 

Therefore  change the repository to `https//cowboysysop.github.io/charts/` in the `Chart.yaml` file  will not work

keitaf replied:
> use the vpa chart from repository https://cowboy.github.io/charts like the metric-server does?

Yes, I think we can directly use the VPA chart as follows:
1. the dashbase chart creates a Secret with TLS certs if `vertical-pod-autoscaler.enabled` is `true`. Its name can be hard-coded to something like `vertical-pod-autoscaler-admission-controller-tls`.
2. In `dashbase/values.yaml`, set `vertical-pod-autoscaler.admissionController.tls.existingSecret` to `vertical-pod-autoscaler-admission-controller-tls`. Then the VPA chart will use the Secret created in the above step. The VPA chart won't create another Secret if `admissionController.tls.existingSecret` is specified ([ref](https://github.com/cowboysysop/charts/blob/master/charts/vertical-pod-autoscaler/templates/admission-controller/tls-secret.yaml))

Can you try and see if the above approach works? I think you can migrate the changes in `dashbase/templates/_helpers.tpl` to `dashbase/templates/vpa_tls_secret.yaml` and also simplify `dashbase/templates/vpa_tls_secret.yaml`.

rluidash replied:
yes,  I think this is possible.. I am testing this approach.. thx

rluidash commented:
Based on the Keita's feedback.. updated this PR 

VPA chart is using repo from `https://cowboy.github.io/charts ` ; the folder  vertocal-pod-autoscaler is removed and the Chart.yaml file is modified with 

```
- name: vertical-pod-autoscaler
  version: 3.0.0 # chart version only, vpa version is 0.9.0
  repository: https://cowboysysop.github.io/charts/
  condition: vertical-pod-autoscaler.enabled
```

And files being modified in dashbase helm chart are below
```
	modified:   dashbase/Chart.lock
	modified:   dashbase/Chart.yaml
	modified:   dashbase/templates/vpa_tls_secret.yaml
	modified:   dashbase/values.yaml
```
dashbase/_helpers.tpl changes are removed  and vpa_tls_secret.yaml file is simplified. 
And by default VPA is not enabled. And can be turned on from dashbase-values.yaml file 

Limitation:
vpa admission controller tls secret name is hardcoded to be  `vertical-pod-autoscaler-admission-controller-tls`

Test result is positive with fresh install and upgrade; against  official [example hamster-vpa](https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/examples/hamster.yaml)

Attached the helm install output for this test install. 
[helm_install_output_dashbase_vpa_12152020.txt](https://github.com/dashbase/dashbase-k8s/files/5699673/helm_install_output_dashbase_vpa_12152020.txt)

keitaf commented on dashbase/templates/vpa_tls_secret.yaml:
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ (printf "%s-%s-%s" "vertical-pod-autoscaler" "admission-controller" "tls") }}
```
nit: I believe the name will always be `vertical-pod-autoscaler-admission-controller-tls`, so we can hard-coded it here.

keitaf commented on dashbase/templates/vpa_tls_secret.yaml:
```diff
@@ -0,0 +1,16 @@
+metadata:
+  name: {{ (printf "%s-%s-%s" "vertical-pod-autoscaler" "admission-controller" "tls") }}
+  labels:
+    {{- include "vertical-pod-autoscaler.admissionController.labels" . | nindent 4 }}
```
I feel we don't need to put those labels, but we can leave it as long as they don't cause any issue.
